testpilot-metrics.js

```diff
@@ -0,0 +1,274 @@
+   * for easy reporting later. e.g. `mainmenu`
+   * @param {string} [$0.study] - Name of a multivariate or A/B test. e.g.
+   * `button-color-test`. If study is present, variant must also be.
+   * @param {string} [$0.variant] - Name of the multivariate test group the
+   * user belongs to. e.g. `green-button`. If variant is present, study must
+   * also be.
+   */
+  sendEvent: function({event, object, category='interactions', study=null, variant=null}) {
+    this._log(`sendEvent called with event = ${event}, object = ${object}, category = ${category}, study = ${study}, variant = ${variant}.`);
+    if (!event) {
+      this._error(`'event' property is required.`);
+      return;
+    } else if (!object) {
+      this._error(`'object' property is required.`);
+      return;
+    }
+
+    this._sendToClient(event, object);
+    if (this.tid) {
+      this._sendToGA(event, object, category, study, variant);
+    }
+  },
+
+  /**
+   * Sends an event to the Mozilla data pipeline via the Test Pilot add-on.
+   * Uses BroadcastChannel for WebExtensions, and nsIObserverService for other
+   * add-on types.
+   * @private
+   * @param {string} event - see `sendEvent` docs
+   * @param {string} object - see `sendEvent` docs
+   */
+  _sendToClient: function(event, object) {
+    const clientEvent = {
+      id: this.id,
+      event: event,
+      object: object
+    };
+
+    if (this.type === 'webextension') {
+      this._channel.postMessage(clientEvent);
+      this._log(`Sent client message via postMessage: ${clientEvent}`);
+    } else {
+      let stringified;
+
+      try {
+        stringified = JSON.stringify(clientEvent);
+      } catch(ex) {
+        this._error(`Unable to serialize metrics event: ${ex}`);
+        return;
+      }
+
+      const subject = {
+        wrappedJSObject: {
+          observersModuleSubjectWrapper: true,
+          object: this.id
+        }
+      };
+
+      try {
+        Services.obs.notifyObservers(subject, 'testpilot::send-metric', stringified);
+        this._log(`Sent client message via nsIObserverService: ${stringified}`);
+      } catch (ex) {
+        this._error(`Failed to send nsIObserver client ping: ${ex}`);
+        return;
+      }
+    }
+  },
+
+  /**
+   * Sends an event to Google Analytics, using sendBeacon and the Measurement
+   * Protocol API.
+   * @private
+   * @param {string} event - see `sendEvent` docs
+   * @param {string} object - see `sendEvent` docs
+   * @param {string} category - see `sendEvent` docs. Note that `category` is
+   * required here, assuming the default value was filled in by `sendEvent`.
+   * @param {string} [study] - see `sendEvent` docs
+   * @param {string} [variant] - see `sendEvent` docs
+   */
+  _sendToGA: function(event, object, category, study, variant) {
+    const data = {
+      v: 1,
+      tid: this.tid,
+      cid: this.cid,
+      t: 'event',
+      ec: category,
+      ea: object,
```

Comment:
This mapping seems a little confusing. In the GA docs and in the realtime view, this `ea` param is called "Event Action" - but it's mapped from `object`. And the `el` is "Event Label" - mapped from `event`.

Confused me a little bit, because I was trying to send events like this:
```js
{ event: 'panel-opened', object: 'snooze-tabs' }
{ event: 'snoozed', object: 'snooze-tabs' }
{ event: 'woke', object: 'snooze-tabs' }
```
And all I was getting were `snooze-tabs` under "Event Action" - and "Event Label" is not even listed on the realtime table.